File: Backend/src/main/java/SearchEngine/SearchEngine.java
package SearchEngine;

/* Search Engine for books
   stored csv files
   made by Guzel Garifullina
   for Sweaty Reader project
*/
import Recommender.BookRecommender;
import org.apache.mahout.cf.taste.common.TasteException;
import java.io.*;
import java.util.ArrayList;
import Basic.Book;

public class SearchEngine {
    private String filePath  = "/home/guzel/" +
            "Programming/SweetyReader/Backend/src/main/java/data/";
    private String fileName = "Formated_BX-Books.csv";
    private File file = new File(filePath + fileName);
    private Book getBook(String[] tokens){
        Book book = new Book();
        book.id = Long.parseLong(tokens[0]);
        book.title = tokens[1];
        book.author = tokens[2];
        book.ps = tokens[3];
        book.pm = tokens[4];
        book.pl = tokens[5];
        return book;
    }
    public ArrayList<Book> getBooks (ArrayList<Long> arr) throws IOException {
        ArrayList<Book> books = new ArrayList<Book>();
        BufferedReader br = null;
        int amt = 0;
        try {
            br = new BufferedReader(new FileReader(file));
            String line;
            amt = arr.size();
            while ((line = br.readLine()) != null) {
                String delims = ";";
                String[] tokens = line.split(delims);
                Long index = Long.parseLong(tokens[0]);
                for (Long e : arr){
                    if (e.equals(index)){
                        amt --;
                        books.add(getBook(tokens));
                        if (amt == 0){
                            return books;
                        }
                    }
                }
            }
        }catch (IOException e) {
            System.err.println("Cannot find file");
        } finally {
            if (br != null) br.close();
            /*if (amt != 0  ){
                System.out.println("Can't find some books");
                for (long e : arr ){
                    System.out.print(e);
                    System.out.print("; ");
                }
                System.out.println();
                for (Book e : books ){
                    System.out.print(e.id);
                    System.out.print("; ");
                }
                throw new IOException();
            }*/
            return books;
        }
    }
    public static void main(String[] args) {
        BookRecommender br = new BookRecommender();
        try {
            SearchEngine searchEngine = new SearchEngine();
            ArrayList items  = br.getRecommendations(8);
            if (items == null){
                System.out.println("Recommendation fail");
                return;
            }
            ArrayList<Book> books = searchEngine.getBooks(items);
            System.out.println("Ok");
            System.out.println(books.get(0).author);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: Backend/src/main/java/SearchEngine/SearchEngine.kt
package SearchEngine

/* Search Engine for books
   stored csv files
   made by Guzel Garifullina
   for Sweaty Reader project
*/

import Basic.Book
import Recommender.BookRecommender1
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.util.*

class SearchEngine {
    private val filePath = "/home/guzel/" + "Programming/SweetyReader/Backend/src/main/java/data/"
    private val fileName = "Formated_BX-Books.csv"
    private val file = File(filePath + fileName)
    private fun getBook(tokens: Array<String>): Book {
        val book = Book()
        book.id = java.lang.Long.parseLong(tokens[0])
        book.title = tokens[1]
        book.author = tokens[2]
        book.ps = tokens[3]
        book.pm = tokens[4]
        book.pl = tokens[5]
        return book
    }

    fun getBooks(arr: ArrayList<Long>): ArrayList<Book> {
        val books = ArrayList<Book>()
        var br: BufferedReader? = null
        var amt = 0
        try {
            br = BufferedReader(FileReader(file))
            var line: String
            amt = arr.size
            line = br.readLine()
            while (line != null) {
                val delims = ";"
                val tokens =
                        line.split(delims.toRegex()).
                                dropLastWhile { it.isEmpty() }.toTypedArray()
                val index = java.lang.Long.parseLong(tokens[0])
                for (e in arr) {
                    if (e == index) {
                        amt--
                        books.add(getBook(tokens))
                        if (amt == 0) {
                            return books
                        }
                    }
                }
                line = br.readLine()
            }
        } catch (e: IOException) {
            System.err.println("Cannot find file")
        } finally {
            if (br != null) br.close()
            return books
        }
    }
    companion object {
        @JvmStatic fun main(args: Array<String>) {
            val br = BookRecommender1()
            try {
                val searchEngine = SearchEngine1()
                val items = br.getRecommendations(8)
                if (items == null) {
                    println("Recommendation fail")
                    return
                }
                val books = searchEngine.getBooks(items)
                println("Ok")
                println(books[0].author)
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }
}
